cmd/admin-revoker/main.go

```diff
@@ -0,0 +1,224 @@
+// Copyright 2015 ISRG.  All rights reserved
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package main
+
+import (
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/cactus/go-statsd-client/statsd"
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/codegangsta/cli"
+
+	// Load both drivers to allow configuring either
+	_ "github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/go-sql-driver/mysql"
+	_ "github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/mattn/go-sqlite3"
+
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/core"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/rpc"
+	"github.com/letsencrypt/boulder/sa"
+
+	gorp "github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp.v1"
+)
+
+var reasons map[int]string = map[int]string{
+	0: "unspecified",
+	1: "keyCompromise",
+	2: "cACompromise",
+	3: "affiliationChanged",
+	4: "superseded",
+	5: "cessationOfOperation",
+	6: "certificateHold",
+	// 7 is unused
+	8:  "removeFromCRL", // needed?
+	9:  "privilegeWithdrawn",
+	10: "aAcompromise",
+}
+
+func loadConfig(c *cli.Context) (config cmd.Config, err error) {
+	configFileName := c.GlobalString("config")
+	configJSON, err := ioutil.ReadFile(configFileName)
+	if err != nil {
+		return
+	}
+
+	err = json.Unmarshal(configJSON, &config)
+	return
+}
+
+func setupContext(context *cli.Context) (rpc.CertificateAuthorityClient, *blog.AuditLogger, *gorp.DbMap) {
+	c, err := loadConfig(context)
+	cmd.FailOnError(err, "Failed to load Boulder configuration")
+
+	stats, err := statsd.NewClient(c.Statsd.Server, c.Statsd.Prefix)
+	cmd.FailOnError(err, "Couldn't connect to statsd")
+
+	auditlogger, err := blog.Dial(c.Syslog.Network, c.Syslog.Server, c.Syslog.Tag, stats)
+	cmd.FailOnError(err, "Could not connect to Syslog")
+	blog.SetAuditLogger(auditlogger)
+
+	ch := cmd.AmqpChannel(c.AMQP.Server)
+
+	cac, err := rpc.NewCertificateAuthorityClient(c.AMQP.CA.Client, c.AMQP.CA.Server, ch)
+	cmd.FailOnError(err, "Unable to create CA client")
+
+	dbMap, err := sa.NewDbMap(c.Revoker.DBDriver, c.Revoker.DBName)
+	cmd.FailOnError(err, "Couldn't setup database connection")
+
+	return cac, auditlogger, dbMap
+}
+
+func AddDeniedNames(tx *gorp.Transaction, names []string) (err error) {
+	sort.Strings(names)
+	deniedCSR := &core.DeniedCsr{Names: strings.ToLower(strings.Join(names, ","))}
+
+	err = tx.Insert(deniedCSR)
+	if err != nil {
+		tx.Rollback()
+		return
+	}
+
+	err = tx.Commit()
```

Comment:
This causes a double-commit, because we commit on [line 201](https://github.com/letsencrypt/boulder/pull/223/files#diff-987b0352e2f1f6bf30d87d0c335934c2R201).

I generally don't like processing Rollback or Commit in sub-methods; I think it's clearer if the err objects propagate upward and happen at the same level as the transaction is opened.